### ISSUE TITLE
Pretrained weights aren't producing meaninful results

### ISSUE THREAD

**Issue by lucasdavid** [Burf/ResNeSt-Tensorflow2#3]
Hello again!

I ran into another issue. I installed torch==1.11.0 and managed to get the pretrained weights.
I made two anecdotal tests (using a few images from Pascal VOC 12 and a few single-label images downloaded from the internet). In both tests the network seems to output almost always the same answer (which are not reasonable for the samples at hand).

So I wonder if there are some problem with the way weights are being loaded?

#### Details of how this was done

1. I downloaded a few images from the web:
```py
import requests

SCORE_CAM_SOURCE = 'https://raw.githubusercontent.com/haofanwang/Score-CAM/master/images'
KERAS_VIS_SOURCE = 'https://raw.githubusercontent.com/keisen/tf-keras-vis/master/examples/images'

class config:
    class data:
        image_size = (512, 512)
        path = '/tf/data'
        images = os.path.join(path, 'misc')
        sources = [
            f'{SCORE_CAM_SOURCE}/ILSVRC2012_val_00000073.JPEG',
            f'{SCORE_CAM_SOURCE}/ILSVRC2012_val_00000091.JPEG',
            f'{SCORE_CAM_SOURCE}/ILSVRC2012_val_00000198.JPEG',
            f'{SCORE_CAM_SOURCE}/ILSVRC2012_val_00000476.JPEG',
            f'{SCORE_CAM_SOURCE}/ILSVRC2012_val_00002193.JPEG',

            f'{KERAS_VIS_SOURCE}/goldfish.jpg',
            f'{KERAS_VIS_SOURCE}/bear.jpg',
            f'{KERAS_VIS_SOURCE}/soldiers.jpg',

            'https://3.bp.blogspot.com/-W__wiaHUjwI/Vt3Grd8df0I/AAAAAAAAA78/7xqUNj8ujtY/s400/image02.png',
            'http://www.aviationexplorer.com/Diecast_Airplanes_Aircraft/delta_Airbus_diecast_airplane.jpg',
            'https://www.petcare.com.au/wp-content/uploads/2017/09/Dalmatian-2.jpg',
            'http://sites.psu.edu/siowfa15/wp-content/uploads/sites/29639/2015/10/dogcat.jpg',
            'https://consciouscat.net/wp-content/uploads/2009/08/multiple-cats-300x225.jpg',
            'https://images2.minutemediacdn.com/image/upload/c_crop,h_843,w_1500,x_0,y_78/f_auto,q_auto,w_1100/v1554995977/shape/mentalfloss/iStock-157312120.jpg',
            'http://www.reptilefact.com/wp-content/uploads/2016/08/Images-of-San-Francisco-Garter-Snake.jpg',
        ]

def load_images():
    global ds, class_names, file_names

    os.makedirs(config.data.images, exist_ok=True)

    file_names = [os.path.join(config.data.images, os.path.basename(f)) for f in config.data.sources]
    class_names = [os.path.splitext(os.path.basename(f))[0] for f in config.data.sources]
    
    for s, d in zip(config.data.sources, file_names):
        if not os.path.exists(d):
            download_image(s, d)
        
    ds = tf.data.Dataset.from_tensor_slices((file_names,))
    ds = ds.map(load_image, num_parallel_calls=tf.data.AUTOTUNE).cache()
    ds = ds.apply(tf.data.experimental.ignore_errors(log_warning=True))
    ds = ds.batch(len(file_names))

    return next(iter(ds))

def download_image(url, path):
    r = requests.get(url, allow_redirects=True)
    with open(path, 'wb') as f:
        f.write(r.content)
    return path

@tf.function
def load_image(x):
    x = tf.io.read_file(x)
    x = tf.image.decode_jpeg(x, channels=3)
    x = tf.image.resize(x, config.data.image_size)
    return x

@tf.function
def augment_fn(image, y):
    image = tf.image.random_brightness(image, .2)
    image = tf.image.random_contrast(image, .75, 1.)
    image = tf.image.random_flip_left_right(image)
    image = tf.image.random_flip_up_down(image)

    return image, y

x = load_images()
```
They can be plotted with matplotlib, and seem to have been correctly loaded. They also work perfectly fine with the Xception network.
![image](https://user-images.githubusercontent.com/1595030/165542184-ea7b9d4a-dd56-4591-a8fe-a840fcd18a79.png)

2. I loaded the model and feed forward the images into it:
```py
import resnest

rs269 = resnest.resnest101(input_shape=(512, 512, 3), include_top=True, weights="imagenet")
rs269.trainable = False

outputs = rs269.predict(images / 255.)
```

3. I used the `applications.imagenet_utils.decode_predictions` to get meaninful names:
```py
predictions = tf.keras.applications.imagenet_utils.decode_predictions(outputs, top=3)
print(
    *(
        ' | '.join(f'{name.rjust(15)} {conf:.2%}' for _id, name, conf in p)
        for p in predictions
    ),
    sep='\n'
)
```
```py
            pot 100.00% |  shower_curtain 0.00% |      wall_clock 0.00%
     wall_clock 99.99% |             pot 0.01% |      hair_slide 0.00%
     wall_clock 100.00% |          bubble 0.00% |      hair_slide 0.00%
         bubble 79.11% |             pot 20.89% |      wall_clock 0.00%
            pot 99.57% |      wall_clock 0.42% |      hair_slide 0.00%
            pot 89.29% |  shower_curtain 10.41% |      wall_clock 0.30%
     wall_clock 100.00% |      hair_slide 0.00% |             pot 0.00%
           mask 99.96% |      wall_clock 0.04% |             pot 0.00%
     wall_clock 99.34% |             pot 0.66% |  shower_curtain 0.00%
           mask 100.00% |             pot 0.00% |      wall_clock 0.00%
         bubble 100.00% |      wall_clock 0.00% |            mask 0.00%
```

**Comment by lucasdavid** [comment]
Just a little more info:
the original code (in zhanghang1989's repo) seems to produce reasonable predictions for these same images:
```py
import torch
from resnest.torch import resnest269
net = resnest269(pretrained=True)

xt = torch.from_numpy(tf.transpose(x, (0, 3, 1, 2)).numpy())

outputs = net.forward(xt / 255.)
preds = outputs.detach().numpy()
probs = tf.nn.softmax(preds).numpy()

print(
    *(
        ' | '.join(f'{name.rjust(15)} {conf:.2%}' for _id, name, conf in p)
        for p in tf.keras.applications.imagenet_utils.decode_predictions(probs, top=3)
    ),
    sep='\n'
)
```
```py
           kite 89.65% | sulphur-crested_cockatoo 2.63% |         vulture 0.26%
     white_wolf 92.33% |     timber_wolf 2.41% |      Arctic_fox 0.16%
         bulbul 90.54% |       chickadee 1.95% |     hummingbird 0.33%
         magpie 79.05% |   oystercatcher 7.87% | European_gallinule 1.44%
        obelisk 63.93% |        pedestal 27.19% |          beacon 0.69%
golden_retriever 87.57% |  cocker_spaniel 2.04% | Labrador_retriever 1.03%
       airliner 94.65% |            wing 1.91% |   space_shuttle 0.14%
      dalmatian 93.70% |      Great_Dane 0.14% | German_short-haired_pointer 0.12%
           chow 17.92% |             tub 8.74% |       tiger_cat 7.73%
     spider_web 94.85% |   garden_spider 2.58% |     barn_spider 0.78%
   garter_snake 35.86% |      sidewinder 20.80% |    horned_viper 15.41%
```

**Comment by Burf** [comment]
First, the official implementation is implemented in Torch, and this is reconstructed as similarly as possible based on Tensorflow.
However, since there are some differences in core operations such as batch normalization and floating point operation between Tensorflow and Torch, it is difficult to say that pre-trained weights implemented in Torch works with TensorFlow at 100% performance.
In conclusion, it is recommended to use pre-trained weights for transfer learning or to learn without pre-trained weights.

**Comment by lucasdavid** [comment]
I found the glitch. In the original code, rSoftmax will normalize the radix axis in the softmax operation (axis 1, default in torch):
```py
x = x.view(batch, self.cardinality, self.radix, -1).transpose(1, 2)
x = F.softmax(x, dim=1)
x = x.reshape(batch, -1)
```

In your code, you correctly reshape the information, but applied the softmax over the last axis:
```py
class rSoftMax(tf.keras.layers.Layer):

def __init__(self, filters, radix, group_size, **kwargs):
        super(rSoftMax, self).__init__(**kwargs)

        self.filters = filters
        self.radix = radix
        self.group_size = group_size

        if 1 < radix:
            self.seq1 = tf.keras.layers.Reshape([group_size, radix, filters // group_size])
            self.seq2 = tf.keras.layers.Permute([2, 1, 3])
            self.seq3 = tf.keras.layers.Activation(tf.keras.activations.softmax)
            self.seq4 = tf.keras.layers.Reshape([1, 1, radix * filters])
            self.seq = [self.seq1, self.seq2, self.seq3, self.seq4]
        else:
            self.seq1 = tf.keras.layers.Activation(tf.keras.activations.sigmoid)
            self.seq = [self.seq1]

    def call(self, inputs):
        out = inputs
        for l in self.seq:
            out = l(out)
        return out
```

This can be fixed by replacing seq3 by `tf.keras.layers.Softmax(axis=1)`, which results in similar (not exactly the same) and meaningful outputs:
```py
Burf/ResNeSt-Tensorflow2:

           kite 64.04% | sulphur-crested_cockatoo 22.03% |  American_egret 0.39%
     white_wolf 57.24% |     timber_wolf 7.44% |          kuvasz 4.54%
         bulbul 92.55% |          coucal 0.19% |       partridge 0.19%
          goose 19.71% |     water_ouzel 14.43% |    black_grouse 8.10%
        obelisk 68.44% |        pedestal 16.04% |          beacon 0.81%
golden_retriever 80.00% | Labrador_retriever 4.71% |          kuvasz 0.89%
       airliner 75.89% |            wing 3.42% |     book_jacket 0.82%
      dalmatian 82.37% | German_short-haired_pointer 0.91% |  English_setter 0.66%
           chow 42.86% | German_shepherd 8.56% |            lion 5.91%
     spider_web 62.87% |     barn_spider 18.96% |   garden_spider 3.23%
   garter_snake 19.36% |        envelope 16.61% |          binder 3.64%
```

**Comment by lucasdavid** [comment]
I can PR if you want.

**Comment by Burf** [comment]
I checked, you were right. I made a mistake on the part where I shouldn't have made a mistake. thanks bro.

**Comment by lucasdavid** [comment]
It's all good. We all make mistakes. :-)
